onnxruntime/contrib_ops/webgpu/quantization/matmul_nbits.cc

```diff
@@ -682,6 +233,156 @@ fn dequantize_packed8xU4(packed_value : u32, zero_point : output_element_t, scal
   return Status::OK();
 }
 
+// Apply similar idea with DP4AMatMulNBitsSmallMProgram algorithm.
+Status MatMulNBitsProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  const auto& a = shader.AddInput("input_a", ShaderUsage::UseValueTypeAlias);
+  const auto& b = shader.AddInput("input_b");
+  shader.AddInput("scales_b");
+  if (has_zero_points_) {
+    shader.AddInput("zero_points", ShaderUsage::UseUniform);
+  }
+  shader.AddOutput("output", ShaderUsage::UseElementTypeAlias);
+  const uint32_t components_a = a.NumComponents();
+  const uint32_t components_b = b.NumComponents() / 4;  // b is stored as uint32 which includs 4 uint8.
+  constexpr uint32_t tile_size_k_vec = 16;
+  uint32_t elements_in_vec = components_b * (32 / nbits_);
+  uint32_t tile_k_size = tile_size_k_vec * elements_in_vec;
+  const uint32_t a_length_per_tile = tile_k_size / components_a;
+
+  shader.AdditionalImplementation() << "const a_length_per_tile = " << a_length_per_tile << "u;\n"
+                                    << "const tile_size_k_vec = " << tile_size_k_vec << ";\n"
+                                    << "const tile_size_k = " << tile_k_size << "u;\n"
+                                    << "const tile_size = " << tile_size_ << "u;\n"
+                                    << "const elements_in_vec = " << elements_in_vec << "u;\n"
+                                    << "const sub_tile_count = " << WorkgroupSizeX() / tile_size_k_vec << "u;\n"
+                                    << "const component_a = " << components_a << "u;\n"
+                                    << "const component_b = " << components_b << "u;\n";
+  shader.AdditionalImplementation() << R"ADDNL_FN(
+  // Shared memory
+  var<workgroup> tile_A : array<input_a_value_t, a_length_per_tile>;
+  var<workgroup> inter_results: array<array<output_element_t, tile_size_k_vec>, tile_size>;
+  fn loadSHMA(batch: u32, a_global: u32, kidx_v: u32, col: u32)
+  {
+    let k_offset = kidx_v / component_a + col;
+    if (batch < uniforms.batch_count && k_offset < uniforms.K_of_a) {
+      tile_A[col] = input_a[batch * uniforms.M * uniforms.K_of_a + a_global * uniforms.K_of_a + k_offset];
+    } else {
+      tile_A[col] = input_a_value_t(0);
+    }
+  }
+)ADDNL_FN"
+                                    << ReadZeroPoint(nbits_, has_zero_points_);
+
+  shader.MainFunctionBody() << R"MAIN_FN(
+  let batch = workgroup_idx / (uniforms.M * uniforms.num_N_tile);
+  let a_global = (workgroup_idx / uniforms.num_N_tile) % uniforms.M;
+  let b_global_base = (workgroup_idx % uniforms.num_N_tile) * tile_size;
+
+  let idx = local_idx % tile_size_k_vec;
+  let idy = local_idx / tile_size_k_vec;
+
+  for (var kidx_v = 0u; kidx_v < uniforms.K; kidx_v += tile_size_k)
```

Comment:
confused by  kidx_v += tile_size_k. We can only add vectorized space size to kidx_v, since _v means its an index in vectorized space. 
Should this be += tile_size_k_vec or is the iterator actually kidx and not kidx_v

Reply:
Sorry for the confusion. Should be kidx and not kidx_v.